### PR TITLE
[XenServer/XCP-ng] Sync the 'platform' setting according to the 'cpu.corespersocket' setting

### PR DESCRIPTION
**Pull request by nvazquez**
### Description

The 'platform' setting is retrieved after a VM is stopped with the platform parameters obtained from Xen/XCP for a certain VM. If a user added the setting 'cpu.corespersocket', it was ignored and the setting was never applied as it was mutually exclusive with the platform 'setting'.

This PR fixes the previous behavior, in the following way:
- If 'cpu.corespersocket'=X AND the 'platform' setting does not contain 'cores-per-socket' -> 'cores-per-socket:X' parameter is added to the 'platform' setting which is sent to Xen/XCP
- If 'cpu.corespersocket'=X AND the 'platform' setting contains 'cores-per-socket:Y' -> The 'platform' setting is updated, containing: 'cores-per-socket:X'
- If 'cpu.corespersocket' is not set, nothing changes

Fixes: #5695 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?
- Deploy a VM using an offering with 2 CPUs (use >=2 CPUs for tests)
- Stop VM -> Settings -> Check the 'platform' setting is visible and does not contain the 'cores-per-socket' parameter
- Add the setting 'cpu.corespersocket' = 2 (CPU number must be a multiple of cores per socket)
- Start VM -> Stop VM -> Check the 'platform' setting contains the parameter 'cores-per-socket:2'
- Edit the setting 'cpu.corespersocket' = 1 (CPU number must be a multiple of cores per socket)
- Start VM -> Stop VM -> Check the 'platform' setting contains the parameter 'cores-per-socket:1'

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
2cents - could this VM property/setting bypass what an admin/operator may not intend? For example, if this setting is set on a template (vs global setting if any) and then bypassed by the user - could user benefit from this setting in a way that admin wouldn't intend to be used?

**Comment by nvazquez** [comment]
@rohityadavcloud the 'cores-per-socket' value could be bypassed or modified but it is restricted by the offering, since a VM won't start if the number of CPUs is not a multiple of the cores per socket number. I think perhaps the 'platform' setting could lead to unintended situations as it includes more information, for example:
````
device-model:qemu-upstream-compat;vga:std;videoram:8;apic:true;viridian:false;timeoffset:0;pae:true;cores-per-socket:2;acpi:1;hpet:true;secureboot:false;nx:true
````

I've checked both settings are visible and editable for users, do you think 'platform' should be hidden for non-admins?

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test centos7 xcpng82

**Comment by nvazquez** [comment]
@blueorangutan test centos7 xcpng82

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 xcpng82

**Comment by nvazquez** [comment]
@blueorangutan test centos7 xcpng82

**Comment by nvazquez** [comment]
@blueorangutan test centos7 xcpng82

**Review by DaanHoogland — APPROVED** [review]
clgtm

File: plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java [hunk 1]
```diff
@@ -1899,13 +1900,25 @@ protected void fillHostInfo(final Connection conn, final StartupRoutingCommand c
         }
     }
 
+    protected void syncPlatformAndCoresPerSocketSettings(String coresPerSocket, Map<String, String> platform) {
+        if (coresPerSocket == null) {
```
Comment (sureshanaparti) [inline plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java:1904]: check if `coresPerSocket` is not blank ?

File: plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java [hunk 2]
```diff
@@ -1899,13 +1900,25 @@ protected void fillHostInfo(final Connection conn, final StartupRoutingCommand c
...
+            return;
+        }
+        if (platform.containsKey(PLATFORM_CORES_PER_SOCKET_KEY)) {
```
Comment (sureshanaparti) [inline plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java:1907]: check for platform not null ?

File: plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java (hunk 1 above)
Comment (nvazquez) [reply plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java]: Thanks, refactored both conditions

**Comment by nvazquez** [comment]
Thanks @sureshanaparti - @blueorangutan package

File: plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java [hunk 3]
```diff
@@ -1899,13 +1900,25 @@ protected void fillHostInfo(final Connection conn, final StartupRoutingCommand c
         }
     }
 
+    protected void syncPlatformAndCoresPerSocketSettings(String coresPerSocket, Map<String, String> platform) {
+        if (org.apache.commons.lang3.StringUtils.isBlank(coresPerSocket) || MapUtils.isEmpty(platform)) {
```
Comment (sureshanaparti) [inline plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java:1904]: add to `platform`, even if it doesn't have any elements ?

```suggestion
        if (org.apache.commons.lang3.StringUtils.isBlank(coresPerSocket) || platform = null) {
```

Comment (DaanHoogland) [reply plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java:1904]: : yes but `==` not `=`!

Comment (nvazquez) [reply plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java:1904]: `isEmpty` checks whether the map is null or empty

Comment (nvazquez) [reply plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java:1904]: Oh, got your point, thanks

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 xcpng82

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test centos7 xcpng82

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 xcpng82

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 xcpng82

**Comment by sureshanaparti** [comment]
> Trillian test result (tid-3108) Environment: xcpng82 (x2), Advanced Networking with Mgmt server 7 Total time taken: 46366 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5892-t3108-xcpng82.zip Smoke tests completed. 91 look OK, 1 have errors Only failed tests results shown below:
> 
> Test	Result	Time (s)	Test File
> test_01_sys_vm_start	`Failure`	0.09	test_secondary_storage.py

The failed test ^^^ is unrelated to this PR. It is intermittent and seems to be environment issue (unable to connect to hosts). Noticed this failure in other PR test results (in xcpng82) as well.

**Comment by sureshanaparti** [comment]
Merging this based on the approvals, smoke test results and tests covered by @nvazquez.

**Pull request merged**